### PR TITLE
ISSUE #877: add verifier.BookkeeperVerifier

### PR DESCRIPTION
**Pull request by athanatos**
Plays and verifies a non-determinstic read/write stream against a
bookkeeper cluster.

Master Issue: #877

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java [hunk 1]
```diff
@@ -0,0 +1,692 @@
...
+ * interface. The test state consists of a set of ledgers in 1 of several states:
+ * 1) opening -- waiting for driver to complete open
+ * 2) open -- valid targets for reads and writes
+ * 3) live -- valid targets for reads
+ * 4) deleting
+ * Each ledger moves in sequence through these states.  See startWrite for the
+ * code driving the lifecycle.
+ */
+public class BookkeeperVerifier {
+    private Queue<Exception> errors = new LinkedList<>();
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java]: final?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java [hunk 2]
```diff
@@ -0,0 +1,692 @@
...
+
+    private synchronized boolean checkReturn(long ledgerID, Integer rc) {
+        if (rc != 0) {
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java]: BKException.Code.SUCCESS != rc

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java [hunk 3]
```diff
@@ -0,0 +1,692 @@
...
+ * 2) open -- valid targets for reads and writes
+ * 3) live -- valid targets for reads
+ * 4) deleting
+ * Each ledger moves in sequence through these states.  See startWrite for the
+ * code driving the lifecycle.
+ */
+public class BookkeeperVerifier {
+    private Queue<Exception> errors = new LinkedList<>();
+
+    private synchronized boolean checkReturn(long ledgerID, Integer rc) {
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java]: use `int` if necessary

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java [hunk 4]
```diff
@@ -0,0 +1,692 @@
...
+    private long nextLedger = 0;
+    private long getNextLedgerID() {
+        return nextLedger++;
+    }
+
+    /**
+     * State required to regenerate an entry.
+     */
+    class EntryInfo {
+        private long entryID;
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java]: line 182-183: make them final?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java [hunk 5]
```diff
@@ -0,0 +1,692 @@
...
+    private int outstandingReadCount = 0;
+    private long nextLedger = 0;
+    private long getNextLedgerID() {
+        return nextLedger++;
+    }
+
+    /**
+     * State required to regenerate an entry.
+     */
+    class EntryInfo {
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java]: make this static?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java [hunk 6]
```diff
@@ -0,0 +1,692 @@
...
+     *  |
+     *  E2 -> Random(E2) -> getBuffer()
+     *  |
+     *  E3 -> Random(E3) -> getBuffer()
+     *  |
+     *  E4 -> Random(E4) -> getBuffer()
+     *  |
+     *  ...
+     */
+    class LedgerInfo {
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java]: static?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java [hunk 7]
```diff
@@ -0,0 +1,692 @@
...
+        private long ledgerID;
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java]: line 220 - 227: make some final if possible

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java [hunk 8]
```diff
@@ -0,0 +1,692 @@
...
+         * the original chain.
+         * @param cb Callback to get result of newOnLastOp if called now
+         * @param newOnLastOp Callback to be invoked on the last decReads or closeWrite,
+         *                    should be passed the cb passed in with the final closeWrite
+         *                    or decReads
+         */
+        void onLastOpComplete(
+                Consumer<Integer> cb,
+                Consumer<Consumer<Integer>> newOnLastOp) {
+            assert (onLastOp == null);
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java]: use checkNotNull or java requireNonNull

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java (hunk 5 above)
Comment (athanatos) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java]: Eh, it references targetEntrySize, not worth passing it in.

Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java]: okay I see

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java (hunk 6 above)
Comment (athanatos) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/verifier/BookkeeperVerifier.java]: It allocates EntryInfo's, easier to leave it non-static.

**Pull request closed without merging**